core/src/main/java/org/infinispan/protostream/impl/TagReaderImpl.java

```diff
@@ -898,51 +880,20 @@ String readString() throws IOException {
          if (length < 0) {
             throw log.negativeLength();
          }
-         if (length <= buf.length) {
-            fillBuffer(length);
-            String value = new String(buf, pos, length, UTF8);
-            pos += length;
-            return value;
-         }
-         return new String(readRawBytesLarge(length), UTF8);
+         throw log.messageTruncated();
       }
 
       @Override
       ByteBuffer readRawByteBuffer(int length) throws IOException {
-         if (length <= end - pos && length > 0) {
-            int from = pos;
-            pos += length;
-            return ByteBuffer.wrap(Arrays.copyOfRange(buf, from, pos));
-         }
-         if (length == 0) {
-            return EMPTY_BUFFER;
-         }
-         if (length < 0) {
-            throw log.negativeLength();
-         }
-         if (length <= buf.length) {
-            fillBuffer(length);
-            int from = pos;
-            pos += length;
-            return ByteBuffer.wrap(Arrays.copyOfRange(buf, from, pos));
-         }
-         // TODO [anistor] implement a readRawByteBufferLarge, using off-heap allocation
-         return ByteBuffer.wrap(readRawBytesLarge(length));
+         byte[] bytes = readRawByteArray(length);
```

Comment:
worth checking `length == 0` to avoid creating an empty `ByteBuffer`? 
```java
         if (length == 0) {
            return EMPTY_BUFFER;
         }
```